solr/core/src/java/org/apache/solr/search/BitDocSet.java

```diff
@@ -237,6 +237,150 @@ public BitDocSet clone() {
     return new BitDocSet(bits.clone(), size);
   }
 
+  @Override
+  public Bits getBits(LeafReaderContext context) {
+    if (context.isTopLevel) {
+      return bits;
+    }
+
+    final int base = context.docBase;
+    final int length = context.reader().maxDoc();
+    final FixedBitSet bs = bits;
+
+    return new Bits() {
+      @Override
+      public boolean get(int index) {
+        return bs.get(index + base);
+      }
+
+      @Override
+      public int length() {
+        return length;
+      }
+    };
+  }
+
+  private static final int NO_DOCS_THIS_SEGMENT = -1;
+  private int[] cachedFloorDocs;
```

Comment:
fwiw, I did intend this code to be (and am pretty sure it is) thread-safe. Because `cachedFloorDocs` is privately built in its entirety by a calling thread and set atomically, the worst-case scenario is that multiple threads end up building the same array, and the array that ends up being cached is arbitrary.
Analogous to `cachedOrdIdxMap` in `SortedIntDocSet`, there are four options:
1. accumulate the relevant information during collection and supply it as an arg to the ctor (feasible? given the different ways in which DocSets might be built?)
2. eagerly compute the relevant information in the ctor (unnecessary work for DocSets that are only used at the global level -- i.e., never need to supply a per-segment iterator) 
3. compute (and cache) the relevant information on-demand the first time `DocSet.iterator(LeafReaderContext)` is called
A fourth option (for `BitDocSet` anyway) is "do nothing", but given that DocSets are cached, and the optimization of returning `null` instead of empty iterators can be valuable to consumers, I do think it's worth the extra effort to do _something_ here ...

Reply:
Lets scope this to BitDocSet (not SortedIntDocSet).  If you are thinking of this as a new optimization... then I'm not sure it's worth it.  Maybe?  I guess a benchmark will say.  It adds complexity.  If this is about an optimization, then for BitDocSet I don't think we need the cached floor arcs but maybe just need a leaf ord based bitmap to indicate which leaves are empty.